packages/actions-shared/src/engage/utils/MessageSendPerformer.ts

```diff
@@ -155,66 +106,113 @@ export abstract class MessageSendPerformer<
-          error
-        })
-        throw error
+    // if messageId or recipientId is not available, then don't cache
+    if (!messageId || !recipientId) return this.sendToRecepient(recepient)
+    return await this.getOrAddCache(
+      `${messageId}-${recipientId?.toLowerCase()}`,
+      () => this.sendToRecepient(recepient),
+      {
+        parse: (cachedValue) => {
+          const parsed = CachedValueFactory.fromString(cachedValue)
+          if (parsed instanceof CachedError) {
+            const error = new IntegrationError(parsed.message, parsed.code, parsed.status)
+            error.retry = false
+            return { error }
+          } else if (parsed.type === CachedResponseType.Success) {
+            return { value: { status: parsed.status } }
+          }
+        },
+        stringify: (cacheable) => {
+          if ('error' in cacheable && cacheable.error && !isRetryableError(cacheable.error)) {
+            const errorDetails = getErrorDetails(cacheable.error)
+            if (errorDetails?.status) {
+              return new CachedError(errorDetails.status, errorDetails.message, errorDetails.code).serialize()
+            }
+          } else if ('value' in cacheable && cacheable.value) {
+            return new CachedValue(cacheable.value.status).serialize()
+          }
+        }
       }
-      this.logError(`Failed to read cache for messageId: ${messageId}, recipientId: ${recipientId}`, {
-        error,
-        message: error.message
-      })
-    }
+    )
+  }
 
-    if (cachedResponse?.type === CachedResponseType.Error && 'message' in cachedResponse) {
-      this.logInfo('Cached error found', {
-        message: cachedResponse.message,
-        code: cachedResponse.code,
-        status: cachedResponse.status
-      })
-      this.statsIncr('error_duplicate')
-      const { message, code, status } = cachedResponse
-      const error = new IntegrationError(message, code, status)
-      error.retry = false
-      throw error
-    } else if (cachedResponse?.type === CachedResponseType.Success) {
-      this.logInfo('Cached response found', { status: cachedResponse.status })
-      this.statsIncr('perform_duplicate')
-      return
+  async getOrAddCache<T>(
+    key: string,
+    createValue: () => Promise<T>,
+    serializer: {
+      //if undefined returned, then the value will not be cached
+      stringify: (cacheable: ValueOrError<T>) => string | void
+      //if undefined returned, then the cache is either corrupted or expired and will be re-executed
+      parse: (cachedValue: string) => ValueOrError<T> | void
     }
+  ): Promise<T> {
+    const engageDestinationCache = this.engageDestinationCache
+    if (!engageDestinationCache) return createValue()
+
+    const cacheRead = await getOrCatch(() => engageDestinationCache.getByKey(key))
+    const cacheReadValue = 'value' in cacheRead ? cacheRead.value : undefined
```

Comment:
same here - too much code and complexity for questionable value. increases cognitive load is the price we pay for perfectionism :)
```suggestion
    const cacheReadValue = cacheRead.value
```